### ISSUE TITLE
[otbn,dv] Teach the ISS to match the side-effect behaviour in the RTL and add cover points to check

### ISSUE THREAD

**Issue by rswarbrick** [lowRISC/opentitan#7809]
For all (possible) multi-cycle instructions, we are going to need to update the ISS so that it can expose register updates on cycles other than the first. We then need to reverse engineer the RTL to generate a spec. See [this comment](https://github.com/lowRISC/opentitan/pull/7655#issuecomment-901217329) from Greg explaining how the RTL works at the moment.

I'm a little concerned that we wouldn't have spotted these mismatches without the conversation on that PR, so we should also add some cover points for faults at all possible points in multi-cycle instructions. For example, underflowing `x1` at the start of an `LW` or overflowing it at the end.

> estimate 16
> remaining 2021-09-20 4

**Comment by cdgori** [comment]
Out of an abundance of caution, I'll just double-check from a security standpoint: is it true that these multi-cycle instructions are all fixed in terms of the number of cycles they consume?  (And/or don't have observably-differing-cycle side effects depending on inputs/state?)

**Comment by rswarbrick** [comment]
Yes, except for CSR/WSR instructions that read from RND, where the latency depends on the EDN.

**Comment by cdgori** [comment]
Thanks for confirmation @rswarbrick  - I assume the variability of the RND read is noted/documented somewhere?  The main consumers of RND (keygen functions) should be able to tolerate this quite easily but it's worth making sure it's clear.

**Comment by imphil** [comment]
@cdgori Yes, it's documented at https://docs.opentitan.org/hw/ip/otbn/doc/#csrs and also in the pseudo-code of the CSR access instructions at e.g. https://docs.opentitan.org/hw/ip/otbn/doc/isa/#csrrs

![image](https://user-images.githubusercontent.com/1467123/130666453-3aaa1125-22a4-488e-8c5e-098fb4b6a81c.png)

Let us know if we should make the wording stronger.

**Comment by rswarbrick** [comment]
Ok, I've just spent quite a while doing cartesian products in my head... and I think these are the cases we have to consider. I make 45 of them in total, roughly half of which come from the complicated BN.LID/BN.SID instructions: I guess this is the price we pay for the clever "invalid GPR" behaviour we implemented :-(

LW
--

The possible errors are: underflow call stack, bogus addr, overflow call
stack. You can\'t under- and overflow the call stack in the same
instruction. What\'s more, there isn\'t even an address to check if you
underflow the call stack, so the only combination to check is:

-   LW: Overflow call stack & bogus address

SW
--

The possible errors are: underflow call stack and bogus addr. These can
happen together if the underflow happens for the value to be stored,
rather than the address.

-   SW: Underflow call stack (for stored value) & bogus address

BEQ
---

The possible errors are: underflow call stack, bad insn addr and branch
at end of loop. Possible combinations:

-   BEQ: Underflow call stack & branch at end of loop
-   BEQ: Bad insn addr & branch at end of loop

BNE
---

This is the same as BEQ.

-   BNE: Underflow call stack & branch at end of loop
-   BNE: Bad insn addr & branch at end of loop

JAL
---

The possible errors are: overflow call stack, bad target addr and jump
at end of loop. All 4 combinations are possible:

-   JAL: Overflow call stack & bad target address
-   JAL: Overflow call stack & jump at end of loop
-   JAL: Bad target address & jump at end of loop
-   JAL: Overflow call stack & bad target address & jump at end of loop

JALR
----

The possible errors are: underflow call stack, overflow call stack, bad
target addr and jump at end of loop. If we underflow the call stack, we
can\'t also overflow it, nor do we have a target address. So the only
combination possible there is underflow + jump at end. Otherwise, all 4
other combinations are possible (like JAL).

-   JALR: Underflow call stack & jump at end of loop.
-   JALR: Overflow call stack & bad target address
-   JALR: Overflow call stack & jump at end of loop
-   JALR: Bad target address & jump at end of loop
-   JALR: Overflow call stack & bad target address & jump at end of loop

CSRRS
-----

The possible errors are: underflow call stack and bad CSR. These can
happen together:

-   CSRRS: Underflow call stack & bad CSR

CSRRW
-----

The possible errors are: underflow call stack, overflow call stack, bad
CSR. Since you can\'t under- and overflow the call stack at once, there
are just two combinations:

-   CSRRW: Underflow call stack & bad CSR
-   CSRRW: Overflow call stack & bad CSR

LOOP
----

The possible errors are: underflow call stack, zero loop count, loop at
end. The first two can\'t go together (because if we underflow the call
stack, we have no loop count), so we get two pairs.

-   LOOP: Underflow call stack & loop at end of loop
-   LOOP: Zero loop count & loop at end of loop

LOOPI
-----

The possible errors are: zero loop count and loop at end. These can
happen together:

-   LOOPI: Zero loop count & loop at end of loop

BN.LID
------

The possible errors are: underflow call stack (grs1), underflow call
stack (grd), both increments set, invalid GPR index and bad data addr.

If we underflow the call stack for grs1, there\'s no architectural
address, so that can\'t happen with bad data addr. If we underflow the
call stack for grd, there\'s no GPR index, so that can\'t cause an
invalid GPR index. However, every other combination is possible. Bin
together the two underflows unless it makes a difference to the possible
behaviour and you get the following list:

-   BN.LID: Underflow call stack & double increment
-   BN.LID: Underflow call stack for grs1 & bad GPR idx
-   BN.LID: Underflow call stack for grd & bad address
-   BN.LID: Double increment & bad GPR idx
-   BN.LID: Double increment & bad address
-   BN.LID: Bad GPR idx & bad address
-   BN.LID: Underflow call stack for grs1 & double increment & bad GPR
    idx
-   BN.LID: Underflow call stack for grd & double increment & bad
    address
-   BN.LID: Double increment & bad GPR idx & bad address

BN.SID
------

The possible errors are: underflow call stack (grs1), underflow call
stack (grs2), both increments set, invalid GPR idx (grs2), bad data addr
(grs1).

Underflowing grs1 means you have no data address and underflowing grs2
means no GPR index. But otherwise, the combinations come out as
basically the same as for BN.LID:

-   BN.SID: Underflow call stack & double increment
-   BN.SID: Underflow call stack for grs1 & bad GPR idx
-   BN.SID: Underflow call stack for grs2 & bad address
-   BN.SID: Double increment & bad GPR idx
-   BN.SID: Double increment & bad address
-   BN.SID: Bad GPR idx & bad address
-   BN.SID: Underflow call stack for grs1 & double increment & bad GPR
    idx
-   BN.SID: Underflow call stack for grs2 & double increment & bad
    address
-   BN.SID: Double increment & bad GPR idx & bad address

BN.MOVR
-------

The possible errors are: underflow call stack (grs1), underflow call
stack (grs2), both increments set, invalid GPR idx (\*grs1), invalid GPR
idx (\*grs2).

Again, this has the same structure as BN.LID, but we can probably bin
together some of the bad GPR idx cases to give:

-   BN.MOVR: Underflow call stack & double increment
-   BN.MOVR: Underflow call stack for grs1 & bad GPR idx (\*grs2)
-   BN.MOVR: Underflow call stack for grs2 & bad GPR idx (\*grs1)
-   BN.MOVR: Double increment & bad GPR idx (either or both)
-   BN.MOVR: Underflow call stack for grs1 & double increment & bad GPR
    idx (\*grs2)
-   BN.MOVR: Underflow call stack for grs2 & double increment & bad GPR
    idx (\*grs1)

@gregac, @imphil: Please could you have a look through and see whether you agree with my list? If so, I'll have a think about how best to test this properly.

**Comment by rswarbrick** [comment]
Hand-written tests to trigger each case now written: see https://github.com/lowRISC/opentitan/pull/8160.

**Comment by rswarbrick** [comment]
I've now sorted out all the mismatches in the cases above except for CSRRS/CSRRW where (based on a conversation in Slack), @GregAC is going to look into modifying the RTL to match the ISS.

This issue started off as a question about multi-cycle instructions, but morphed into one about instructions that could raise multiple errors, possibly in a single cycle (oops).

Work still remaining:
- Think carefully about the multi-cycle instructions and add cover points for any interesting error behaviour they should have.
- Add a cover point for each "multi-error" case listed above.
- Work out a good way to run through these multi-error cases (all 46 of them). Maybe a directed test approach would work. Discussion scheduled for this evening's all-team DV meeting.
- Check we can fill the coverage points I've added.
